core/modules/file/file.entity.inc

```diff
@@ -126,6 +126,77 @@ class File extends Entity {
       'options' => array(),
     );
   }
+
+  /**
+   * Overrides Entity::createAccess().
+   */
+  public static function createAccess($bundle = NULL, $account = NULL) {
+    $rights = &backdrop_static(__METHOD__, array());
+
+    // If no user object is supplied, the access check is for the current user.
+    if (empty($account)) {
+      $account = $GLOBALS['user'];
+    }
+
+    // If we've already checked access for this node, user and op, return from
+    // cache.
+    if (isset($rights[$account->uid][$bundle])) {
+      return $rights[$account->uid][$bundle];
+    }
+
+    $rights[$account->uid][$bundle] = user_access('access content', $account);
+    return $rights[$account->uid][$bundle];
+  }
+
+  /**
+   * Overrides Entity::access().
+   *
+   * @param string $op
+   *   The operation to be performed on the node. Possible values are:
+   *   - "update".
+   *   - "delete".
+   * @param User|AnonymousUser|object $account
+   *   (optional) The user to check for. Leave it to NULL to check for the
+   *   global user.
+   *
+   * @return bool
+   *   TRUE if access is granted, FALSE otherwise.
+   */
+  public function access($op, $account = NULL) {
+    // Casting class with private property causes errors due to added prefix.
+    // e.g. "\0" . 'File' . "\0". So use static array instead.
+    $rights = &backdrop_static(__METHOD__, array());
+
+    if (!in_array($op, array('update', 'delete'), TRUE)) {
+      // If the $op was not one of the supported ones, we return access denied.
+      return FALSE;
+    }
+    // If no user object is supplied, the access check is for the current user.
+    if (empty($account)) {
+      $account = $GLOBALS['user'];
+    }
+
+    $cid = $this->id();
+
+    // If we've already checked access for this node, user and op, return from
+    // cache.
+    if (isset($rights[$account->uid][$cid][$op])) {
+      return $rights[$account->uid][$cid][$op];
+    }
+
+    if ($op == 'update') {
+      $rights[$account->uid][$cid][$op] = user_access('manage files', $account);
+      return $rights[$account->uid][$cid][$op];
+    }
+    elseif ($op == 'delete') {
+      $rights[$account->uid][$cid][$op] = user_access('delete files', $account);
+      return $rights[$account->uid][$cid][$op];
+    }
```

Comment:
As `user_access()` calls are already cached, do we need this complicated caching mechanism? It's good for future-proofing, but without any more complicated checks beyond `user_access()` calls, this cache seems unnecessary.